spec/haskell/src/SEL4/Object/SchedContext.lhs

```diff
@@ -723,3 +727,26 @@ This module uses the C preprocessor to select a target architecture.
 >     threadSet (\t -> t { tcbInReleaseQueue = False }) tcbPtr
 >     setReprogramTimer True
 >     return tcbPtr
+
+In preemptible code, the kernel may explicitly mark a preemption point with the "preemptionPoint" function. The preemption will only be taken if an interrupt has occurred and the preemption point has been called "workUnitsLimit" times.
+
+> workUnitsLimit = 0x64
+
+> preemptionPoint :: KernelP ()
+> preemptionPoint = do
+>     lift $ modifyWorkUnits ((+) 1)
+>     workUnits <- lift $ getWorkUnits
+>     when (workUnitsLimit <= workUnits) $ do
+>       lift $ setWorkUnits 0
+>       preempt <- lift $ doMachineOp (getActiveIRQ True)
+>       case preempt of
+>           Just irq -> throwError ()
+>           Nothing -> do
```

Comment:
It's probably not necessary, but I think the following would better match the C and might be a bit easier in crefine. It shouldn't have an impact on any of the other proofs.

```
>       if isJust preempt
>           then throwError ()
>           else do
```